### DISCUSSION TITLE
combinations coverage

### DISCUSSION THREAD

**Discussion by adrian-herscu** [instancio/instancio#786] [Q&A]
### Feature description

For simplicity, let's say we have an object with one boolean field. There can be two possible values: true and false. Let's say that we want to generate at least two instances of such object. 

Is there a way to ensure that all combinations will be generated? 

How about objects with more fields? 

Not sure if this request has a possible implementation. 
Couldn't find any clue about it in current documentation. Maybe it's worth mentioning it? 

```java
    @Test
    public void shouldGenerateAllCombinations() {
        @ToString
        class Foo {
            boolean b;
        }

        Instancio.stream(Foo.class)
            .limit(2)
            .forEach(foo -> log.debug(foo.toString()));
    }
```

(on first attempt got false twice)

**Comment by adrian-herscu** [comment]
Answering my own question:
```java
        Instancio.stream(Foo.class)
            .distinct()
            .limit(2)
            .forEach(foo -> log.debug(foo.toString()));
```
Still must allow enough objects to be generated in order to cover all combinations.
Any other sugesstion?

**Comment by armandino** [comment]
@adrian-herscu thank you for the suggestion, that's an interesting idea. Currently, there's no support for generating combinations. I'll need some time to investigate how feasible it would be to implement such a feature.

One alternative would be to use `@RepeatedTest` to generate a big enough sample size that will cover most of the combinations. This can work well for simple tests.

There are a couple of other libraries that support combinations:

- https://junit-pioneer.org/docs/cartesian-product
- https://github.com/google/TestParameterInjector

**Comment by adrian-herscu** [comment]
I have another use case:

```java
@Test
    public void shouldGenerateAllCombinations2() {
        @ToString
        @EqualsAndHashCode
        class Foo {
            String a;
            String b;
        }

        Instancio.of(Foo.class)
            .withNullable(all(field("a"), field("b")))
            .set(all(field("a"), field("b")), "any")
            .stream()
            .distinct()
            .limit(4) // must generate enough to cover all combinations
            .forEach(foo -> log.debug(foo.toString()));
    }
```

This works as expected:
```
18:55:39.416 [DEBUG] [main ] [1]: InstancioSelfTests - Foo(a=any, b=any)
18:55:39.419 [DEBUG] [main ] [1]: InstancioSelfTests - Foo(a=null, b=null)
18:55:39.419 [DEBUG] [main ] [1]: InstancioSelfTests - Foo(a=any, b=null)
18:55:39.420 [DEBUG] [main ] [1]: InstancioSelfTests - Foo(a=null, b=any)
```

But, the requirement might be: generate a null field *only* once:
| null | any |
| any | null | 

Any suggestion about how to achieve that?

**Comment by armandino** [comment]
One option is to simply loop through individual selectors and set it to null. Here's an example:

```java
@Test
public void shouldGenerateAllCombinations() {
    @Data class Foo { String a, b, c; }

    Selector[] nullableFields = {
            field(Foo::getA),
            field(Foo::getB),
            field(Foo::getC)
    };

    Arrays.stream(nullableFields)
            .map(nullField -> createWithSingleNullField(Foo.class, nullableFields, nullField))
            .forEach(foo -> System.out.println(foo.toString()));
}

static <T> T createWithSingleNullField(Class<T> klass, Selector[] nullableFields, Selector nullField) {
    return Instancio.of(klass)
             // order matters
            .set(all(nullableFields), "any") // (1) set all fields to "any"
            .set(nullField, null) // (2) overwrite this one to null
            .create();
}

// Output:
// Foo(a=null, b=any, c=any)
// Foo(a=any, b=null, c=any)
// Foo(a=any, b=any, c=null)
```

**Comment by adrian-herscu** [comment]
Added `javax.annotation.Nullable` to all nullable fields, then tried:
```java
@UtilityClass
public class SelectEx extends Select {
    public static PredicateSelector allNullables() {
        return fields(f -> nonNull(f
            .getDeclaredAnnotation(Nullable.class)));
    }
}
```
but `Select` is final. Now I have to import both `Select` and `SelectEx`.

**Comment by adrian-herscu** [comment]
Thought about some generic solution. Tried to mark every nullable field with `javax.annotation.Nullable`, then defined:

```java
    public static PredicateSelector allNullables(
        final Class<?> declaringClass) {
        return fields(
            f -> nonNull(f.getDeclaredAnnotation(Nullable.class)));
    }
```

and used it like this:

```java
public class Address {
    @Builder.Default
    @Nullable
    public String line1     = "line1";
    ... other fields ...
    public static <T extends Address> Model<T> model(final Class<T> clazz) {
        return Instancio.of(clazz)
            .withNullable(allNullables(Address.class))
            ... other operations ....
            .toModel();
```

It works ok with the Address class, but fails with its subclass (while calling `toModel`) -- the reason of the failure is due to `ReflectionUtils` class:
```java
    public static Field getField(final Class<?> klass, final String fieldName) {
        try {
            return klass.getDeclaredField(Verify.notNull(fieldName, "null field name"));
        } catch (NoSuchFieldException ex) {
            throw Fail.withUsageError("invalid field '" + fieldName + "' for " + klass, ex);
        } catch (SecurityException ex) {
            throw new InstancioException("Unable to access '" + fieldName + "' of " + klass, ex);
        }
    }
```
If it called `getField` instead of `getDeclaredField` then it would work.


(related to #774 - reusing models)

**Comment by armandino** [comment]
@adrian-herscu could you provide the subclass definition and a sample that produces the error? I'll try to reproduce on my end to see what's going on.

**Reply by adrian-herscu** [reply]
Was a bit bussy, sorry for late reply.

```java
    public static PredicateSelector allNullables(
        final Class<?> declaringClass) {
        return fields(
            f -> declaringClass.equals(f.getDeclaringClass())
                && nonNull(f.getDeclaredAnnotation(Nullable.class)));
    }
```

fixed the issue
